src/org/zalando/stups/pierone/api_v2.clj

```diff
@@ -197,19 +207,24 @@
                    (ring/header "Content-Length" size)
                    ; layers are already GZIP compressed!
                    (ring/header "Content-Encoding" "identity")))
-    (resp "image not found" request :status 404)))
+    (resp {:errors [(assoc-in (:BLOB_UNKNOWN errors) [:detail] {"Digest" digest})]} request :status 404)))
 
 (defn read-manifest
   "Read manifest JSON and throw 'Bad Request' if invalid"
   [data]
-  (try
-    (json/parse-string (slurp data))
-    (catch Exception e
-      (api/throw-error 400 "invalid manifest JSON"))))
+  (if (map? data)
+    data
+    (try
+      (walk/keywordize-keys (json/parse-string (slurp data)))
+      (catch Exception e
+        (api/throw-error 400 (str data))))))
 
 (defn get-fs-layers
    [manifest]
-   (map #(get % "blobSum") (get manifest "fsLayers")))
+   (if (= 1 (get manifest :schemaVersion))
```

Comment:
Also we should probably explicitly check for `(= 2 schemaVersion)` instead of assuming it? And throw an error if it's not 2.